edn/src/types_pp.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![cfg_attr(feature = "cargo-clippy", allow(linkedlist))]
+
+use std::collections::{BTreeSet, BTreeMap, LinkedList};
+use std::borrow::Cow;
+use pp::{PrettyPrintTree, PrintInstr};
+use types::Value;
+
+// A pretty printing implementation for an `edn::Value`, focused on limited
+// whitespace expansion and readability for mentat queries.
+impl<'a> From<&'a Value> for PrettyPrintTree<'a> {
+    fn from(src: &Value) -> PrettyPrintTree {
+        match *src {
+            Value::Vector(ref v) => v.into(),
+            Value::List(ref v) => v.into(),
+            Value::Set(ref v) => v.into(),
+            Value::Map(ref v) => v.into(),
+            Value::Text(ref v) => string_leaf!(format!("\"{}\"", v)),
+            _ => string_leaf!(src.to_string())
+        }
+    }
+}
+
+impl<'a> From<&'a Vec<Value>> for PrettyPrintTree<'a> {
```

Comment:
I don't think there's any necessary abstraction leak here. Everything can indeed be inlined instead.

None of these subsequent implementation blocks are actually required, I've opted for implementing multiple `From` traits because it seemed more idiomatic, and gave us more pretty printable data structures for free. Splitting things into smaller functions also avoids cyclomatic complexity and is (maybe subjectively) easier to read. 

> We want to be able to print edn::Value, not arbitrary vectors (or maps, or sets) of values.

Unless there's some non-written requirement about allowing anything other than a single `edn::Value` to be pretty printable, I don't see the harm in having pretty printing implemented for free for multiple kinds of source types with this approach. They might be useful especially in tests, when having to compare different kinds of data. I don't see what is gained by not having them and inlining everything.

Like I said above, changing this is easy, but I'd like to understand why first.